File: Difficulty: Medium/Next Greater Element/next-greater-element.py
# User function Template for python3

class Solution:
    # Function to find the next greater element for each element of the array.
    def nextLargerElement(self, arr):
        ans=[-1]*len(arr)
        st=[]
        
        for i in range(len(arr)):
            # print(ans,st)
            while st and arr[st[-1]]<arr[i]:# when a big number greater than stack 
            # top is come than pop the stack till top of stack is less
                ans[st.pop()]=arr[i]
             
            st.append(i)
        return ans    





#{ 
 # Driver Code Starts
# Initial Template for Python 3

t = int(input())  # number of test cases
for _ in range(t):
    arr = list(map(int, input().split()))  # input array
    s = Solution().nextLargerElement(arr)  # find the next greater elements

    # Output formatting
    if s:
        print(" ".join(map(str, s)))  # Print next greater elements
    else:
        print("[]")  # Print empty list if no next greater element is found
    print("~")
# } Driver Code Ends

File: Difficulty: Medium/Next Greater Element/next-greater-element.java
class Solution {
    public ArrayList<Integer> nextLargerElement(int[] arr) {
        // code here
        ArrayList<Integer>ans=new ArrayList<>();
        int n=arr.length;
        Stack<Integer>st=new Stack<>();

        for(int i=n-1;i>=0;i--){
            while(!st.isEmpty()&& st.peek()<=arr[i]) st.pop();
            
            if(!st.isEmpty()&& st.peek()>arr[i]) ans.add(st.peek());
            else ans.add(-1);
            
            st.push(arr[i]);
        }
        Collections.reverse(ans);
        return ans;
    }
}